collections/src/main/scala/strawman/collection/immutable/LongMap.scala

```diff
@@ -0,0 +1,442 @@
+      if ((this.left eq left) && (this.right eq right)) this.asInstanceOf[LongMap.Bin[S]]
+      else LongMap.Bin[S](prefix, mask, left, right)
+    }
+  }
+}
+
+// Iterator over a non-empty LongMap.
+private[immutable] abstract class LongMapIterator[V, T](it: LongMap[V]) extends AbstractIterator[T] {
+
+  // Basically this uses a simple stack to emulate conversion over the tree. However
+  // because we know that Longs are only 64 bits we can have at most 64 LongMap.Bins and
+  // one LongMap.Tip sitting on the tree at any point. Therefore we know the maximum stack
+  // depth is 65
+  var index = 0
+  var buffer = new Array[AnyRef](65)
+
+  def pop() = {
+    index -= 1
+    buffer(index).asInstanceOf[LongMap[V]]
+  }
+
+  def push(x: LongMap[V]): Unit = {
+    buffer(index) = x.asInstanceOf[AnyRef]
+    index += 1
+  }
+  push(it)
+
+  /**
+    * What value do we assign to a tip?
+    */
+  def valueOf(tip: LongMap.Tip[V]): T
+
+  def hasNext = index != 0
+  final def next(): T =
+    pop() match {
+      case LongMap.Bin(_,_, t@LongMap.Tip(_, _), right) => {
+        push(right)
+        valueOf(t)
+      }
+      case LongMap.Bin(_, _, left, right) => {
+        push(right)
+        push(left)
+        next()
+      }
+      case t@LongMap.Tip(_, _) => valueOf(t)
+      // This should never happen. We don't allow LongMap.Nil in subtrees of the LongMap
+      // and don't return an LongMapIterator for LongMap.Nil.
+      case LongMap.Nil => throw new IllegalStateException("Empty maps not allowed as subtrees")
+    }
+}
+
+private[immutable] class LongMapEntryIterator[V](it: LongMap[V]) extends LongMapIterator[V, (Long, V)](it){
+  def valueOf(tip: LongMap.Tip[V]) = (tip.key, tip.value)
+}
+
+private[immutable] class LongMapValueIterator[V](it: LongMap[V]) extends LongMapIterator[V, V](it){
+  def valueOf(tip: LongMap.Tip[V]) = tip.value
+}
+
+private[immutable] class LongMapKeyIterator[V](it: LongMap[V]) extends LongMapIterator[V, Long](it){
+  def valueOf(tip: LongMap.Tip[V]) = tip.key
+}
+
+/**
+  *  Specialised immutable map structure for long keys, based on
+  *  <a href="http://citeseer.ist.psu.edu/okasaki98fast.html">Fast Mergeable Long Maps</a>
+  *  by Okasaki and Gill. Essentially a trie based on binary digits of the integers.
+  *
+  *  Note: This class is as of 2.8 largely superseded by HashMap.
+  *
+  *  @tparam T      type of the values associated with the long keys.
+  *
+  *  @since 2.7
+  *  @define Coll `immutable.LongMap`
+  *  @define coll immutable long integer map
+  *  @define mayNotTerminateInf
+  *  @define willNotTerminateInf
+  */
+sealed abstract class LongMap[+T] extends Map[Long, T]
+  with MapOps[Long, T, Map, LongMap[T]]
+  with StrictOptimizedIterableOps[(Long, T), Iterable, LongMap[T]] {
+
+  protected[this] def fromSpecificIterable(coll: strawman.collection.Iterable[(Long, T)]): LongMap[T] = {
+    //TODO should this be the default implementation of this method in StrictOptimizedIterableOps?
```

Comment:
Yes, probably, that’s a good point. I’ve noticed that we could probably provide more default implementations of the factory methods in the `Ops` traits in general. I’ve created #335 to keep track of that.